kernel/src/grant.rs

```diff
@@ -142,24 +144,30 @@ impl<T: Default> Grant<T> {
     }
 
     pub fn grant(&self, appid: AppId) -> Option<AppliedGrant<T>> {
-        unsafe {
-            appid.kernel.process_map_or(None, appid, |process| {
-                if let Some(grant_ptr_ref) = process.grant_ptr(self.grant_num) {
-                    let cntr = *(grant_ptr_ref as *mut *mut T);
-                    if cntr.is_null() {
+        appid.kernel.process_map_or(None, appid, |process| {
+            if let Some(grant_ptr) = process.grant_ptr(self.grant_num) {
+                // Check if this grant has been allocated yet
+                let grant = {
+                    if grant_ptr.is_null() {
```

Comment:
Here we do a null check, so we shouldn't need to call `new_unchecked` afterwards. The whole block of code can be simplified as follows, without any `unsafe`.

```rust
NonNull::new(grant_ptr)
  .cast::<T>()
  .map(|grant| {
    AppliedGrant {
      appid: appid,
      grant: grant,
      _phantom: PhantomData,
    }
})
```